freqtrade/exchange/__init__.py

```diff
@@ -105,132 +94,231 @@ def validate_pairs(pairs: List[str]) -> None:
+    except ccxt.InsufficientFunds as e:
+        raise DependencyException(
+            'Insufficient funds to create limit sell order on market {}.'
+            'Tried to sell amount {} at rate {} (total {}).'
+            'Message: {}'.format(pair, amount, rate, rate*amount, e)
+        )
+    except ccxt.InvalidOrder as e:
+        raise DependencyException(
+            'Could not create limit sell order on market {}.'
+            'Tried to sell amount {} at rate {} (total {}).'
+            'Message: {}'.format(pair, amount, rate, rate*amount, e)
+        )
+    except ccxt.NetworkError as e:
+        raise NetworkException(
+            'Could not place sell order due to networking error. Message: {}'.format(e)
+        )
+    except ccxt.BaseError as e:
+        raise OperationalException(e)
 
 
 def get_balance(currency: str) -> float:
     if _CONF['dry_run']:
         return 999.9
 
-    return _API.fetch_balance()[currency]
+    # ccxt exception is already handled by get_balances
+    balances = get_balances()
+    return balances[currency]['free']
 
 
-def get_balances():
+def get_balances() -> dict:
     if _CONF['dry_run']:
-        return []
+        return {}
+
+    try:
+        balances = _API.fetch_balance()
+        # Remove additional info from ccxt results
+        balances.pop("info", None)
+        balances.pop("free", None)
+        balances.pop("total", None)
+        balances.pop("used", None)
+
+        return balances
+    except ccxt.NetworkError as e:
+        raise NetworkException(
+            'Could not get balance due to networking error. Message: {}'.format(e)
+        )
+    except ccxt.BaseError as e:
+        raise OperationalException(e)
 
-    return _API.fetch_balance()
 
-# @cached(TTLCache(maxsize=100, ttl=30))
 @retrier
 def get_ticker(pair: str, refresh: Optional[bool] = True) -> dict:
-    return _API.fetch_ticker(pair)
+    global _TICKER_CACHE
+    try:
+        if not refresh:
+            if _TICKER_CACHE and pair in _TICKER_CACHE:
+                return _TICKER_CACHE[pair]
+        _TICKER_CACHE[pair] = _API.fetch_ticker(pair)
+        return _TICKER_CACHE[pair]
+    except ccxt.NetworkError as e:
+        raise NetworkException(
+            'Could not load tickers due to networking error. Message: {}'.format(e)
+        )
+    except ccxt.BaseError as e:
+        raise OperationalException(e)
 
 
-# @cached(TTLCache(maxsize=100, ttl=30))
 @retrier
-def get_ticker_history(pair: str, tick_interval) -> List[List]:
-    # TODO: tickers need to be in format 1m,5m
-    # fetch_ohlcv returns an [[datetime,o,h,l,c,v]]
+def get_ticker_history(pair: str, tick_interval: str) -> List[Dict]:
     if 'fetchOHLCV' not in _API.has or not _API.has['fetchOHLCV']:
-        logger.warning('Exhange %s does not support fetching historical candlestick data.',
-                       _API.name)
-        return []
+        raise OperationalException(
```

Comment:
I think we shouldn't raise Exceptions here, we can log it and return `[]`. It's usually easier to handle empty result than try to get the exception handling right.

Reply:
I think we should raise exceptions. It plays nicer with the retrier decorator to repeat failed requests if it's a NetworkException, and prevents confusing it with just empty results. It also feels more pythonic to handle exceptions rather than having to validate the data returned from the API once more.

Having multiple similar exception handling clauses might look a bit ugly but it's probably not much more work than writing a general exception handler. It also allows for granularity in how we want to handle an exception. We probably want to handle a failed call to `buy` differently than a failed call to `get_ticker_history`, which we can control by which exception we raise.

Reply:
@enenn +1